### PR TITLE
Feature/delete users folders

### LINKED ISSUE #152: Delete all objects associated with user on user delete
#### Description

When a user is deleted, delete all associate objects and folders expect those that are already published.

#### DoD (Definition of Done)

both objects and users are delete from DB

#### Testing

Unit tests and Integration tests

### LINKED ISSUE #160: Move drafts in folders to drafts in user when folder published
#### Description

Currently the action is to delete drafts but we should keep them for the user

#### DoD (Definition of Done)

Drafts are move from folders to user drafts when folder publish

#### Testing

unit and integration tests

### PR DESCRIPTION
**Pull request by blankdots**
### Related issues

closes #152 
closes #160 

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Changes Made

1. for delete users, also delete unpublished folders and associated objects
2. when delete folders also delete objects
3. fix typos
4. introduce a more complex structure to user drafts so that they can be easier retrieved.

### Testing

- [x] Unit Tests
- [x] Integration Tests

### CODE REVIEW

File: metadata_backend/api/handlers.py [hunk 1]
```diff
@@ -696,13 +709,27 @@ async def delete_user(self, req: Request) -> Response:
...
-        user = await operator.delete_user(current_user)
-        LOG.info(f"DELETE user with ID {user} was successful.")
+        user = await operator.read_user(current_user)
+
+        for fold in user["folders"]:
+            _folder = await fold_ops.read_folder(fold)
+            if not _folder["published"]:
+                for obj in _folder["drafts"] + _folder["metadataObjects"]:
+                    await obj_ops.delete_metadata_object(obj["schema"], obj["accessionId"])
+                await fold_ops.delete_folder(fold)
```
Comment (vvaltia) [inline metadata_backend/api/handlers.py:726]: Worth considering if `fold` and `_folder` are distinguishable enough from each other

File: metadata_backend/api/operators.py [hunk 2]
```diff
@@ -755,27 +752,29 @@ async def remove_object(self, folder_id: str, collection: str, accession_id: str
...
+                LOG.error(reason)
+                raise web.HTTPBadRequest(reason=reason)
+            if not delete_success:
+                reason = f"Deleting for {folder_id} from database failed."
+                LOG.error(reason)
+                raise web.HTTPBadRequest(reason=reason)
+            else:
+                LOG.info(f"Deleting folder with id {folder_id} to database succeeded.")
+                return folder_id
+        return None
```
Comment (vvaltia) [inline metadata_backend/api/operators.py:777]: Maybe this makes sense, but should/could the if-else be inside `try` ?

**Review by teemukataja — APPROVED** [review]

File: tests/test_operators.py [hunk 3]
```diff
@@ -744,34 +740,30 @@ async def test_folder_update_fails_with_bad_patch(self):
...
         operator.db_service.remove.assert_called_once()
         self.assertEqual(len(operator.db_service.remove.mock_calls), 1)
 
     async def test_folder_object_remove_fails(self):
         """Test folder remove object fails."""
         operator = FolderOperator(self.client)
-        operator.db_service.exists.side_effect = ConnectionFailure
+        operator.db_service.remove.side_effect = ConnectionFailure
         with self.assertRaises(HTTPBadRequest):
             await operator.remove_object(self.test_folder, "study", self.accession_id)
```
Comment (vvaltia) [inline tests/test_operators.py:769]: Just a general comment: test, test2, patch, patch2 etc are not the most descriptive for variables or values

**Review by vvaltia — APPROVED** [review]

File: metadata_backend/api/operators.py (hunk 2 above)
Comment (blankdots) [reply metadata_backend/api/operators.py]: good point, fixed

File: metadata_backend/api/handlers.py (hunk 1 above)
Comment (blankdots) [reply metadata_backend/api/handlers.py]: good point, fixed

**Review by vvaltia — APPROVED** [review]

File: tests/test_operators.py (hunk 3 above)
Comment (blankdots) [reply tests/test_operators.py:769]: variable names and function descriptions improved, so that they make more sense now.

**Pull request merged**